finetuner/tuner/callback/model_checkpoint.py

```diff
@@ -0,0 +1,141 @@
+        the end of every epoch, or after a fixed number of training batches.
+    """
+
+    def __init__(
+        self,
+        filepath: str = None,
+        save_best_only: Optional[bool] = False,
+        monitor: Optional[str] = 'val_loss',
+        mode: Optional[str] = 'auto',
+    ):
+        """
+        :param filepath: string or `PathLike`, path to save the model file.
+        :param save_best_only: if `save_best_only=True` only the best model
+        model will be saved according to the quantity monitored
+        :param monitor: if `monitor='loss'` best bodel saved will be according
+        to the training loss, if `monitor='val_loss'` best model saved will be
+        according to the validation loss
+        :param mode: one of {'auto', 'min', 'max'}. If `save_best_only=True`, the
+        decision to overwrite the current save file is made based on either
+        the maximization or the minimization of the monitored quantity.
+        For `val_acc`, this should be `max`, for `val_loss` this should be
+        `min`, etc. In `auto` mode, the mode is set to `max` if the quantities
+        monitored are 'acc' or start with 'fmeasure' and are set to `min` for
+        the rest of the quantities.
+        """
+        self.__logger = JinaLogger(self.__class__.__name__)
+        self.__filepath = filepath
+        self.__save_best_only = save_best_only
+        self.__monitor = monitor
+        if not filepath:
+            raise ValueError(
+                '``filepath`` parameter is mandatory. Pass it in parameters'
+            )
+
+        if mode not in ['auto', 'min', 'max']:
+            self.__logger.warning(
+                'ModelCheckpoint mode %s is unknown, ' 'fallback to auto mode.', mode
+            )
+            mode = 'auto'
+
+        if mode == 'min':
+            self.__monitor_op = np.less
+            self.__best = np.Inf
+        elif mode == 'max':
+            self.__monitor_op = np.greater
+            self.__best = -np.Inf
+        else:
+            if 'acc' in self.__monitor or self.__monitor.startswith('fmeasure'):
+                self.__monitor_op = np.greater
+                self.__best = -np.Inf
+            else:
+                self.__monitor_op = np.less
+                self.__best = np.Inf
+
+    def get_monitor_op(self):
+        return self.__monitor_op
+
+    def get_best(self):
+        return self.__best
+
+    def on_train_epoch_end(self, tuner: 'BaseTuner'):
+        """
+        Called at the end of the training epoch.
+        """
+        if self.__monitor == 'loss':
+            self._save_model(tuner)
+
+    def on_val_end(self, tuner: 'BaseTuner'):
+        """
+        Called at the end of the validation epoch.
+        """
+        if self.__monitor == 'val_loss':
+            self._save_model(tuner)
+
+    def _save_model(self, tuner):
+        if self.__save_best_only:
+            current = tuner.state.current_loss
```

Comment:
One thing: the loss is only given as the loss per batch - but if you want to use `val_loss` what you need to do is to compute the average of it. To do this you'll need to hook into `on_val_batch_end` and collect the losses there

Reply:
So if I understand correctly the loss received in the `tuner state` in `on_epoch_end` and `on_val_end` is the loss of the last batch not the whole data ?